File: Sorting/QuickSort/QuickSort.cs
//Author: Tushar Jaiswal
//Creation Date: 07/26/2018

using System;

namespace ConsoleApp
{
	public class Solution
	{
		static void Main(string[] args)
		{
			int[] arr = { 2, 5, 7, 3, 7, 2, 1, 8, -3 };
			Sorting.QuickSort(arr, 0, arr.Length - 1);
			foreach (int i in arr)
			{
				Console.Write(i + " ");
			}
		}
	}

	public class Sorting
	{
		/// <summary>
		/// This quick sort function sorts an input array of integers in increasing order.
		/// </summary>
		/// <param name="arr">the input array of integers</param>
		public static void QuickSort(int[] arr, int low, int high)
		{
			if (low < high)
			{
				int pivot = Partition(arr, low, high);

				QuickSort(arr, low, pivot - 1);
				QuickSort(arr, pivot + 1, high);
			}
		}

		private static int Partition(int[] arr, int low, int high)
		{
			Random rnd = new Random();
			int pivot = rnd.Next(low, high + 1);
			Swap(ref arr[pivot], ref arr[high]);
			int i = low - 1;

			for(int j = low; j < high; j++)
			{
				if(arr[j] <= arr[high])
				{
					i++;
					Swap(ref arr[i], ref arr[j]);
				}
			}
			Swap(ref arr[i + 1], ref arr[high]);
			return (i + 1); 
		}

		/// <summary>
		/// This function swaps the values of its parameters which are passed in by reference. It is a generic method and it can operate on different types.
		/// </summary>
		/// <param name="a">First parameter</param>
		/// <param name="b">Second parameter</param>
		public static void Swap<T>(ref T a, ref T b)
		{
			T temp;
			temp = a;
			a = b;
			b = temp;
		}
	}
}

File: Sorting/QuickSort/QuickSort.java
/**
* Execution class that illustrates the use of the sorting methods.
* @author  Tushar Jaiswal
*/

import java.util.concurrent.ThreadLocalRandom;

class Solution 
{
    public static void main(String[] args) 
    {
        int[] arr = {2,5,7,3,7,2,1,8,-3};
        Sorting.quickSort(arr, 0, arr.length - 1);
        for (int i : arr) 
        {
            System.out.print(i + " ");
        }
    }
}

/**
* Class with different sorting methods.
* @author  Tushar Jaiswal
*/
class Sorting 
{
    /**
    * This insertion sort function sorts an input array of integers in increasing order.
    * @param arr The input array of integers.
    * @param low Lower bound of the array to operate on
    * @param high Higher bound of the array to operate on
    * @return Nothing.
    */
    public static void quickSort(int[] arr, int low, int high) 
    {
        if (low < high)
        {
            int pivot = partition(arr, low, high);

            quickSort(arr, low, pivot - 1);
            quickSort(arr, pivot + 1, high);
        }
    }
    
    /**
    * This function generates a random pivot and arranges all elements lesser than or equal to itself before its position and all elements greater than itself after its position.
    * @param arr The input array of integers.
    * @param low Lower bound of the array to operate on.
    * @param high Higher bound of the array to operate on.
    * @return The index of the pivot element.
    */
    private static int partition(int[] arr, int low, int high)
    {
        int pivot = ThreadLocalRandom.current().nextInt(low, high + 1);
        
        int temp = arr[pivot];
        arr[pivot] = arr[high];
        arr[high] = temp;
        
        int i = low - 1;

        for(int j = low; j < high; j++)
        {
            if(arr[j] <= arr[high])
            {
                i++;
                temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        return (i + 1); 
    }
}